### PR TITLE
Nikita/apriltags

### PR DESCRIPTION
**Pull request by mishra-nikita**

### CODE REVIEW

File: autovalet_gazebo/launch/aruco_sim.launch [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        <!-- <remap from="/camera_info" to="/stereo/$(arg eye)/camera_info" /> -->
+        <remap from="/camera_info" to="/realsense/frontCamera/color/camera_info" />
+        <!-- <remap from="/image" to="/stereo/$(arg eye)/image_rect_color" /> -->
+        <remap from="/image" to="/realsense/frontCamera/color/image_raw" />
+
+        <param name="image_is_rectified" value="True"/>
+        <param name="marker_size"        value="$(arg markerSize)"/>
+        <param name="marker_id"          value="$(arg markerId)"/>
+        <param name="reference_frame"    value="$(arg ref_frame)"/>   <!-- frame in which the marker pose will be refered -->
+        <!-- <param name="camera_frame"       value="stereo_gazebo_$(arg eye)_camera_optical_frame"/> -->
```
Comment (kob51) [inline autovalet_gazebo/launch/aruco_sim.launch]: remove this line

File: autovalet_husky/launch/aruco_harware.launch [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        <!-- <remap from="/camera_info" to="/stereo/$(arg eye)/camera_info" /> -->
+        <remap from="/camera_info" to="/frontCamera/color/camera_info" />
+        <!-- <remap from="/image" to="/stereo/$(arg eye)/image_rect_color" /> -->
+        <remap from="/image" to="/frontCamera/color/image_raw" />
+
+        <param name="image_is_rectified" value="True"/>
+        <param name="marker_size"        value="$(arg markerSize)"/>
+        <param name="marker_id"          value="$(arg markerId)"/>
+        <param name="reference_frame"    value="$(arg ref_frame)"/>   <!-- frame in which the marker pose will be refered -->
+        <!-- <param name="camera_frame"       value="stereo_gazebo_$(arg eye)_camera_optical_frame"/> -->
```
Comment (kob51) [inline autovalet_husky/launch/aruco_harware.launch]: delete if not beign used

File: autovalet_gazebo/scripts/validate_arucotags.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        """
+        To ping the first message of gt and populate the world to pose offset
+        """
+        # self.isFirstCb = True
+        self.offset = None
+        self.positionErrorLog = []
+        self.yawErrorLog = []
+        self.runningAverage = []
+
+    def gt_cb(self, gt):
```
Comment (kob51) [inline autovalet_gazebo/scripts/validate_arucotags.py:38]: more verbose fxn names? `ground_truth_callback`?

File: autovalet_gazebo/scripts/validate_arucotags.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        """
+
+        #TODO: get list index from name of first message
+        self.gt = gt.pose[-7] ## 7th (from last) name 'aruco_visual_marker_7::marker'
+        self.offs = gt.pose[-5] #ground truth published by gazebo link state /::base_link w.r.t the gazebo world origin
+        self.offset = np.array([self.offs.position.x,  self.offs.position.y, self.offs.position.z, self.quat_to_euler(self.offs.orientation)[2]])
+
+    def est_cb(self, est):
+        """
+        pose: estimated pose published by est_topic (in our case: EKF localization's filtered odometry)
```
Comment (kob51) [inline autovalet_gazebo/scripts/validate_arucotags.py]: get rid of this comment, we're not using EKF, right?

File: autovalet_gazebo/scripts/validate_arucotags.py (hunk 3 above)
Comment (mishra-nikita) [reply autovalet_gazebo/scripts/validate_arucotags.py:38]: Let this be the way it is, the function description makes it clearer. 
I kept it this way so that it stays consistent with the other validation scripts.

Cleaned up the rest

File: autovalet_arucotags/aruco_gazebo/.gitignore [hunk 5]
```diff
@@ -0,0 +1,2 @@
+#Temp
+*~
```
Comment (evamo0508) [inline autovalet_arucotags/aruco_gazebo/.gitignore]: NIT: remove this file since it is not necessary once we deleted the `.git` folder

File: autovalet_gazebo/config/husky_gazebo_lot.rviz [hunk 6]
```diff
@@ -438,6 +438,6 @@ Window Geometry:
     collapsed: false
   Views:
     collapsed: false
-  Width: 1600
-  X: 0
+  Width: 928
+  X: 992
   Y: 24
```
Comment (evamo0508) [inline autovalet_gazebo/config/husky_gazebo_lot.rviz]: NIT: just want to start a discussion on whether we should push RVIZ changes to `master` every time we modified it according to our own needs

File: autovalet_gazebo/launch/aruco_sim.launch [hunk 7]
```diff
@@ -0,0 +1,32 @@
+<!-- To run simulation correctly along with this launch file:
+Download the package aruco_gazebo for the markers from here: https://github.com/joselusl/aruco_gazebo.git
+We import aruco_visual_marker_7 model in our world
+Detection done using aruco_ros package from pal-robotics: https://github.com/pal-robotics/aruco_ros.git 
+Put the packages in ~catkin_ws/src/-->
```
Comment (evamo0508) [inline autovalet_gazebo/launch/aruco_sim.launch]: are those info still legit for us? if we're not using the `aruco_visual_marker_7` model, we should either remove those descriptions or revise it accordingly to avoid future confusion.

File: autovalet_gazebo/launch/aruco_sim.launch [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    <arg name="markerSize"      default="0.034"/>    <!-- in m -->
+    <arg name="eye"             default="left"/>
+    <arg name="marker_frame"    default="aruco_marker_frame"/>
+    <arg name="ref_frame"       default=""/>  <!-- leave empty and the pose will be published wrt param parent_name -->
+    <arg name="corner_refinement" default="LINES" /> <!-- NONE, HARRIS, LINES, SUBPIX -->
+
+
+    <node pkg="aruco_ros" type="single" name="aruco_single">
+        <remap from="/camera_info" to="/realsense/frontCamera/color/camera_info" />
+        <remap from="/image" to="/realsense/frontCamera/color/image_raw" />
```
Comment (evamo0508) [inline autovalet_gazebo/launch/aruco_sim.launch]: Now that @SubramanianKrish 's MR has been merged, should those topic names be without `/realsense`?

File: autovalet_gazebo/launch/parking_lot.launch [hunk 9]
```diff
@@ -31,9 +31,7 @@ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSI
     <arg name="simulation" value="true"/>
   </include>
 
-
-  <arg name="world_file" default="cmu_parking_lot.world"/>
-
+  <arg name="world_file" default="aruco_vert_dumpster.world"/>
```
Comment (evamo0508) [inline autovalet_gazebo/launch/parking_lot.launch]: yea revert this change.

File: autovalet_gazebo/scripts/validate_arucotags.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+        self.gt = None
+        self.estimate = None
+        # self.last_gt = None
+        # self.filename = file
+
+        """
+        To ping the first message of gt and populate the world to pose offset
+        """
+        # self.isFirstCb = True
```
Comment (evamo0508) [inline autovalet_gazebo/scripts/validate_arucotags.py]: remove unused comments to keep the code clean?

File: autovalet_gazebo/scripts/validate_arucotags.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+        self.positionErrorLog = []
+        self.yawErrorLog = []
+        self.runningAverage = []
+
+    def gt_cb(self, gt):
+        """
+        localization validation callback function
+        """
+
+        #TODO: get list index from name of first message
```
Comment (evamo0508) [inline autovalet_gazebo/scripts/validate_arucotags.py:43]: is this still a `TODO`?

File: autovalet_gazebo/scripts/validate_arucotags.py [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    rospy.init_node("aruco_tags_validation", anonymous=False)
+    
+    ground_truth_topic = '/gazebo/link_states'
+    estimate_topic = '/aruco_single/pose'
+    # sample_dist = 0 #0.3
+    # outFile = "/home/nikita/catkin_ws_av/src/autovalet/autovalet_gazebo/data/"
+    noise = 0.008
+    # outFile = outFile + str(noise)
+
+    errorHandler = aruco_tags_validation(ground_truth_topic, estimate_topic) #, sample_dist, outFile)
```
Comment (evamo0508) [inline autovalet_gazebo/scripts/validate_arucotags.py]: also here, clean up the comments if they are not useful anymore

File: autovalet_arucotags/aruco_gazebo/.gitignore (hunk 5 above)
Comment (mishra-nikita) [reply autovalet_arucotags/aruco_gazebo/.gitignore]: Done

File: autovalet_gazebo/config/husky_gazebo_lot.rviz (hunk 6 above)
Comment (mishra-nikita) [reply autovalet_gazebo/config/husky_gazebo_lot.rviz]: should I keep this change or undo it?
I don't feel too strongly towards either decision.

File: autovalet_gazebo/launch/aruco_sim.launch (hunk 7 above)
Comment (mishra-nikita) [reply autovalet_gazebo/launch/aruco_sim.launch]: They are up-to-date. But this stuff is already in the readme, so I can remove it from here.

File: autovalet_gazebo/launch/aruco_sim.launch (hunk 8 above)
Comment (mishra-nikita) [reply autovalet_gazebo/launch/aruco_sim.launch]: I will rebase

File: autovalet_gazebo/launch/parking_lot.launch (hunk 9 above)
Comment (mishra-nikita) [reply autovalet_gazebo/launch/parking_lot.launch]: done

File: autovalet_gazebo/scripts/validate_arucotags.py (hunk 11 above)
Comment (mishra-nikita) [reply autovalet_gazebo/scripts/validate_arucotags.py:43]: yes

File: autovalet_gazebo/scripts/validate_arucotags.py (hunk 10 above)
Comment (mishra-nikita) [reply autovalet_gazebo/scripts/validate_arucotags.py]: done

File: autovalet_gazebo/scripts/validate_arucotags.py (hunk 12 above)
Comment (mishra-nikita) [reply autovalet_gazebo/scripts/validate_arucotags.py]: done

File: autovalet_gazebo/config/husky_gazebo_lot.rviz (hunk 6 above)
Comment (evamo0508) [reply autovalet_gazebo/config/husky_gazebo_lot.rviz]: @kob51  @SubramanianKrish

**Comment by SubramanianKrish** [comment]
Could you fix the aruco_harware launch typo? Thanks!

**Comment by mishra-nikita** [comment]
> Could you fix the aruco_harware launch typo? Thanks!

Done

**Review by SubramanianKrish — DISMISSED** [review]
